streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -77,10 +77,10 @@ Segment getOrCreateSegment(final long segmentId, final ProcessorContext context)
             if (!isSegment(segment, segmentId)) {
                 cleanup(segmentId);
             }
-            if (!segments.containsKey(key)) {
-                Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment previousSegment = segments.putIfAbsent(key, newSegment);
+            if (previousSegment == null) {
```

Comment:
You can also replace `return segments.get(key);` with `return previousSegment == null ? newSegment : previousSegment`.